plutus-core/plutus-ir/test/GeneratorSpec/Terms.hs

```diff
@@ -0,0 +1,182 @@
+        go _ tmCtx tm [] = (tmCtx, tm)
+        go i tmCtx tm (InstApp ty : insts) = go i tmCtx (TyInst () tm ty) insts
+        go i tmCtx tm (InstArg ty : insts) = go (succ i) (Map.insert y ty tmCtx)
+                                                         (Apply () tm (Var () y)) insts
+          where y = Name "y" i
+
+-- | Check what's in the leaves of the generated data
+prop_stats_leaves :: Property
+prop_stats_leaves =
+  -- No shrinking here because we are only collecting stats
+  forAllDoc "_,tm" genTypeAndTerm_ (const []) $ \ (_, tm) ->
+  tabulate "leaves" (map (filter isAlpha . show . prettyPirReadable) $ leaves tm) $ property True
+  where
+    -- Figure out what's at the leaves of the AST,
+    -- including variable names, error, and builtins.
+    leaves (Var _ x)        = [x]
+    leaves (TyInst _ a _)   = leaves a
+    leaves (Let _ _ _ b)    = leaves b
+    leaves (LamAbs _ _ _ b) = leaves b
+    leaves (Apply _ a b)    = leaves a ++ leaves b
+    leaves Error{}          = [Name "error" $ toEnum 0]
+    leaves Builtin{}        = [Name "builtin" $ toEnum 0]
+    leaves _                = []
+
+-- | Check the ratio of duplicate shrinks
+prop_stats_numShrink :: Property
+prop_stats_numShrink =
+  -- No shrinking here because we are only collecting stats
+  forAllDoc "ty,tm" genTypeAndTerm_ (const []) $ \ (ty, tm) ->
+  let shrinks = map snd $ shrinkClosedTypedTerm (ty, tm)
+      n = length shrinks
+      u = length $ nubHashableOn showRenameTerm shrinks
+      r | n > 0     = 5 * ((n - u) * 20 `div` n)
+        | otherwise = 0
+  in tabulate "distribution | duplicates" ["         | " ++ show r ++ "%"] True
+
+-- | Specific test that `inhabitType` returns well-typed things
+prop_inhabited :: Property
+prop_inhabited =
+  -- No shrinking here because if the generator
+  -- generates nonsense shrinking will be nonsense.
+  forAllDoc "ty,tm" (genInhab mempty) (const []) $ \ (ty, tm) -> typeCheckTerm tm ty
+  where
+    -- Generate some datatypes and then immediately call
+    -- `inhabitType` to test `inhabitType` directly instead
+    -- of through the whole term generator. Quick-ish way
+    -- of debugging "clever hacks" in `inhabitType`.
+    genInhab ctx = runGenTm $ local (\ e -> e { geTypes = ctx }) $
+      genDatatypeLets $ \ dats -> do
+        ty <- genType $ Type ()
+        tm <- inhabitType ty
+        return (ty, foldr (\ dat -> Let () NonRec (DatatypeBind () dat :| [])) tm dats)
+
+-- | Check that there are no one-step shrink loops
+prop_noTermShrinkLoops :: Property
+prop_noTermShrinkLoops =
+  -- Note that we need to remove x from the shrinks of x here because
+  -- a counterexample to this property is otherwise guaranteed to
+  -- go into a shrink loop.
+  forAllDoc "ty,tm" genTypeAndTerm_
+    (\(ty', tm') -> filter ((/= tm') . snd) $ shrinkClosedTypedTerm (ty', tm')) $ \(ty, tm) ->
+  tm `notElem` map snd (shrinkClosedTypedTerm (ty, tm))
```

Comment:
cc @MaximilianAlgehed , any thoughts?

Maybe we can just turn the number of tests way down for these tests?

Reply:
I'd be tempted to change the property to
```
  tm `notElem` take 100_000 (map snd $ shrinkClosedTypedTerm (ty, tm))
```
(or some other reasonable number) to avoid it blowing up.

The problem of getting a reasonable order of shrinkings is harder, of course. Maybe good practice would be to interleave shrinkings of subcomponents instead of appending them. In any case, QuickCheck is in desperate need of a better language for expressing shrinking.